### DISCUSSION TITLE
In multisite setup, the error pages in XMC are not rendered for each site.

### DISCUSSION THREAD

**Discussion by isyedakhtar** [Sitecore/jss#1622] [Help]
### Description

In (headless) SXA, we can define the error pages either on the Sitegrouping or Settings item itself. XM Foundation repo code is pulling the layout from edge for those error pages and renders them. For example in getStaticProps of 404.tsx 
`export const getStaticProps: GetStaticProps = async (context) => {
  const site = siteResolver.getByName(config.jssAppName);
  const errorPagesService = new GraphQLErrorPagesService({
    endpoint: config.graphQLEndpoint,
    apiKey: config.sitecoreApiKey,
    siteName: site.name,
    language: context.locale || config.defaultLanguage,
  });
}`
When getting the site information, we are using config.jssAppName. It works well for a single website hosted. If we are using the multisite plugin and have a single project, every site would render the error pages defined for the website that has a presence in config.jssAppName i.e. package.json app name.

Can someone suggest, what's the best way of tackling this. I thought to resolve it by removing the getStaticProps and use the Sitecore context within the main function. That could be a quick resolution but no SSG for a static page. 

### Additional information

I couldn't find a way to create a discussion under the repo https://github.com/sitecorelabs/xmcloud-foundation-head so thought to start a thread here.

**Comment by illiakovalenko** [comment]
@isyedakhtar Hey, I think it's possible to create an issue ticket there: https://github.com/sitecorelabs/xmcloud-foundation-head-dev/issues
But here it's fine.
I can suggest you take a look at [site](https://github.com/sitecorelabs/xmcloud-foundation-head-dev/blob/main/src/sxastarter/src/lib/page-props-factory/plugins/site.ts) plugin. It extracts a real requested site, you can use this code and do the same in _getStaticProps_ and use the value
```ts
export const getStaticProps: GetStaticProps = async (context) => {
  const path =
      context.params === undefined
        ? '/'
        : Array.isArray(context.params.path)
        ? context.params.path.join('/')
        : context.params.path ?? '/';

  // Get site name (from path)
  const siteData = getSiteRewriteData(path, config.jssAppName);

  // Resolve site by name
  const site = siteResolver.getByName(siteData.siteName);

  const errorPagesService = new GraphQLErrorPagesService({
    endpoint: config.graphQLEndpoint,
    apiKey: config.sitecoreApiKey,
    siteName: site.name,
    language: context.locale || config.defaultLanguage,
  });
  let resultErrorPages: ErrorPages | null = null;

  if (!process.env.DISABLE_SSG_FETCH) {
    try {
      resultErrorPages = await errorPagesService.fetchErrorPages();
    } catch (error) {
      console.log('Error occurred while fetching error pages');
      console.log(error);
    }
  }

  return {
    props: {
      headLinks: [],
      layoutData: resultErrorPages?.notFoundPage?.rendered || null,
    },
  };
};
```

**Reply by isyedakhtar** [reply]
Hi @illiakovalenko that site plugin and the code you shared has the same problem. They are using config.jssAppName to fetch the site information. In multisite setup, the config.jssAppName will not work as it will only be pointing to one of the sites. Which means, for all the sites, we are serving 404 page of the site config.jssAppName

**Reply by illiakovalenko** [reply]
@isyedakhtar Hi. Nope, `getSiteRewriteData(path, config.jssAppName);` - here is _config.jssAppName_ is a fallback value, you will get siteName from the rewrite path

**Reply by isyedakhtar** [reply]
@illiakovalenko  The page is generated statically. getStaticProps is called only once for all the sites in a multisite environment. When I tried this change, I see that it is generating static pages for each site. In my case there two sites 'basic' and 'skaters'. It generates static pages for both basic/404 and then skaters/404. However, because the SiteData is only being created once and it points to Skater site, both error pages are taken from skater site content.

**Reply by art-alexeyenko** [reply]
> However, because the SiteData is only being created once and it points to Skater site, both error pages are taken from skater site content.

@isyedakhtar can you share your implementation for 404.tsx? 404 page will be generated for both sites, and getStaticProps should execute for both, with different site names.  Can you verify if that's the case on your end?

**Reply by isyedakhtar** [reply]
@art-alexeyenko The 404 is statically generated for both sites as the page is part of the XM content home/404. When there is 404 error, the 404.tsx is executed from the next code which then calls the getStaticProps inside it. At this point, the getStaticProps doesn't know the context site. I tried using getSiteRewriteData as @illiakovalenko  mentioned above, but the context.params always returns undefined. This 
makes it fallback to the jssAppName.

![image](https://github.com/Sitecore/jss/assets/10513890/cb156e1c-48fb-4060-b14e-cdef48102e99)

**Comment by illiakovalenko** [comment]
@matkovskyi Can you, please, take a look a this issue and consider implementing it if it's possible? Seems like it's a limitation for multisite solutions using ErrorPages, isn't it?

**Reply by sc-ruslanmatkovskyi** [reply]
@illiakovalenko We have a task for implementing ErrorPages with multisite supporting. But now it isn't technically possible as I know

**Comment by ambrauer** [comment]
Next.js (Pages Router) only allows static 404/500 error pages, and they are compiled during build time. If you have i18n configured, those locales will be taken into account (and static versions of each language are built). This works because i18n is a built-in feature of Next.js. However, multisite is not and relies on dynamic site resolution and custom rewrite path segments to be used during build time.

This is a [known limitation for JSS multisite](https://doc.sitecore.com/xmc/en/developers/jss/216/jss-xmc/the-next-js-multisite-add-on.html#:~:text=404%20and%20500%20custom%20error%20pages%20are%20not%20currently%20supported.%20Instead%2C%20use%20the%20default%20error%20pages.). The starter code uses your configured _default_ site for the 404/500 page content.
**If you need unique 404/500 pages per site, then you can always serve sites using a dedicated rendering host (a single site)** using the same codebase.

Note this is not a problem on Next.js App Router (which we will provide support for in the future). Even the official Next.js multi-tenant example did not have [multi-tenant error pages](https://github.com/vercel/platforms/blob/main/app/%5Bdomain%5D/%5Bslug%5D/not-found.tsx#L6-L10) until App Router 😉.
